### ISSUE TITLE
[BUG] error when handle types defined with lambda

### ISSUE THREAD

**Issue by steve02081504** [doxygen/doxygen#10313]
**Describe the bug**
Doxygen doesn't seem to be able to handle types defined with lambda correctly.
e.g.
```c++
template<size_t size>
using unsigned_specific_size_t=decltype(lambda{
	#define TYPE_MAPPER(type) if constexpr(size == sizeof(type))return (type)0;else
	#line 1 "https://github.com/ELC-lang/ELC/tree/master/parts/header_file/files/elc/../../../_share/basic_environment/arithmetic_mapper/unsigned_mapper.hpp"
//unsigned_mapper.hpp
TYPE_MAPPER(uint8_t)
TYPE_MAPPER(uint16_t)
TYPE_MAPPER(uint32_t)
TYPE_MAPPER(uint64_t)
#if defined(ELC_BASE_ENV_HAS_INT128)
TYPE_MAPPER(uint128_t)
#endif
TYPE_MAPPER(::std::uintmax_t)
TYPE_MAPPER(size_t)
TYPE_MAPPER(unsigned char)
TYPE_MAPPER(char8_t)
TYPE_MAPPER(char16_t)
TYPE_MAPPER(char32_t)
TYPE_MAPPER(wchar_t)
TYPE_MAPPER(unsigned short)
TYPE_MAPPER(unsigned int)
TYPE_MAPPER(unsigned long)
TYPE_MAPPER(unsigned long long)

//file_end

	#line 105 "https://github.com/ELC-lang/ELC/tree/master/parts/header_file/files/elc/../../../_share/basic_environment/_body.hpp"
	#undef TYPE_MAPPER
	{}
}());

```

**Expected behavior**
Handle it correctly.

**Screenshots**
https://elc-lang.github.io/ELC/namespaceelc_1_1defs_1_1basic__environment.html#a526ab5592597e881c00a5d308b568bd0
![图片](https://github.com/doxygen/doxygen/assets/31927825/89b9b075-3efc-4fbd-9ae5-d7caea9376cf)

**To Reproduce**
[docs.zip](https://github.com/doxygen/doxygen/files/12641917/docs.zip)

**Version**
Doxygen version used: 1.9.8
github action with https://github.com/mattnotmitt/doxygen-action/commit/452281fb0f5f8a3e8d3f7fa5359cb58f249dfa28

**Stack trace**
none.

**Additional context**
none.

**Comment by albert-github** [comment]
The example in the attachment (docs.zip) is a bit large and the problem there is hard to pinpoint, though the small example in the issue already shows the problem of the fact tat the `using` / `decltype` /`lambda` is not handled properly. The problem is even more notably when changing:
```
#define TYPE_MAPPER(type) if constexpr(size == sizeof(type))return (type)0;else
```
into
```
#define TYPE_MAPPER(type) if (constexpr(size == sizeof(type)))return (type)0;else
```


(the problem is independent of using clang)

**Comment by steve02081504** [comment]
> https://github.com/doxygen/doxygen/issues/10313#issuecomment-1722425140

@albert-github 
Thanks for the additions, I think I might have to add a little more stuff here
In the example code snippet provided, `lambda` is a macro that embellishes the c++23 code rather than a keyword, and you can get it like this (sorry this was an oversight on my part!)
```c++
#define lambda []
```
https://en.cppreference.com/w/cpp/language/lambda

Another thing is that I think `if constexpr(...)` is a syntactically similar keyword to `if(...)` is a syntactically close keyword, whereas `if (constexpr(...))` is likely to get an error in the compiler, which may be something to watch out for!
https://en.cppreference.com/w/cpp/language/if

**Comment by doxygen** [comment]
@steve02081504 Please verify if [the latest commit](https://github.com/doxygen/doxygen/commit/a928976086e6d9038b206ea12d42751204fa1cc0) fixes the issue for you. Do not close the issue, this will be done automatically when the next release is made available.

**Comment by albert-github** [comment]
@doxygen Output looks better though:

When taking the small example in the issue (and `MACRO_EXPANSION=NO`): D:\doxygen_bugs\_Triage\issues_10\issue_10313\ex_small I get:
![image](https://github.com/doxygen/doxygen/assets/5223533/d0ed5ea3-9842-45eb-8ea2-b84ac4374195)

and
![image](https://github.com/doxygen/doxygen/assets/5223533/86b0c996-1db2-436e-9218-3e0a0ac15432)

- strange that in the brief description the `TYPE_MAPPER` lines are not mentioned and only the brackets, though this is easily explained as `SKIP_FUNCTION_MACROS=YES` is set by default, so a bit understandable.
- the "initialization" is a multi line initialization and should probably be handled as such, now it is squashed into 1 line.

When setting ``MACRO_EXPANSION=YES` and `SKIP_FUNCTION_MACROS=NO` we get for the detailed part:
![image](https://github.com/doxygen/doxygen/assets/5223533/cda5d76f-5fcb-47f6-b20d-92b0c419561f)

**Comment by steve02081504** [comment]
@doxygen Thanks for your help, it looks good now!
I've been modifying my project while you've been improving doxygen. I've reverted [a workground for msvc as it was fixed](https://developercommunity.visualstudio.com/t/10311041), and unfortunately this new modification has led to another strange output from doxygen.
The new problematic code looks like this:
```c++
//warp for not_unreadable_istream to make it unreadable
template<class stream_T> requires not_unreadable_istream_class<stream_T>
struct unreadable_wrap:stream_T,instance_struct<unreadable_wrap<stream_T>>,virtual decltype(lambda{
	if constexpr(noexcept_stream_class<stream_T>) {
		if constexpr(data_stream_class<stream_T>)
			return type_info<noexcept_data_istream_t>;
		else
			return type_info<noexcept_text_istream_t<typename stream_T::char_type>>;
	}
	else {
		if constexpr(data_stream_class<stream_T>)
			return type_info<data_istream_t>;
		else
			return type_info<text_istream_t<typename stream_T::char_type>>;
	}
}())::type{
	//coool stuffs!
};
```
Screenshot of the generated output:
![图片](https://github.com/doxygen/doxygen/assets/31927825/344883de-9b9f-4d25-a382-0fdce5837899)
New reproducible code:
[docs.zip](https://github.com/doxygen/doxygen/files/12706349/docs.zip)
The path to the problematic output:
`output/structelc_1_1defs_1_1stream__n_1_1unreadable__wrap.html#a958334cb0bfb618ab93c797a08cebedf`

**Comment by doxygen** [comment]
@steve02081504 I'm not sure, I even understand how to read the code fragment you provided. 
Can you isolate it in a small self-contained example (i.e. one that compiles)?

**Comment by doxygen** [comment]
@steve02081504 Please verify if the latest commit fixes the problem (with the limitations mentioned).

**Comment by albert-github** [comment]
@doxygen though less important, but what is the status of the output issue mentioned in https://github.com/doxygen/doxygen/issues/10313#issuecomment-1732255366

**Comment by steve02081504** [comment]
> @steve02081504 Please verify if the latest commit fixes the problem (with the limitations mentioned).

It's working fine thank you for your help! (and apologies for my late testing)
I can't wait to see the next version of doxygenXD
By the way, could you please take a look at https://github.com/doxygen/doxygen/issues/9621 sometime? I'm not sure my wishes were made clear and understood in that issue but it looks like it's been stuck there for a while.

**Comment by doxygen** [comment]
> @doxygen though less important, but what is the status of the output issue mentioned in [#10313 (comment)](https://github.com/doxygen/doxygen/issues/10313#issuecomment-1732255366)

I think the "multi-line initializer for using" can indeed better be handled as in other places. Can you make a new issue for this specifically?

**Comment by albert-github** [comment]
The requested (https://github.com/doxygen/doxygen/issues/10313#issuecomment-1741817360) issue:  "Multi line initialization for using" #10334

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.10.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
